### PR TITLE
Remote Config added

### PR DESCRIPTION
**Pull request by kayahanbaskeser**
Remote Config added

### CODE REVIEW

**Comment by m0skit0** [comment]
Please use C# naming conventions, methods start with uppercase, thanks.

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+        public void applyDefault(Dictionary<String, System.Object> map) => Call("applyDefault", DictionaryOperations.DictionaryToAndroidJavaObject(map));
+
+        public void applyDefault(string xmlPath) => Call("applyDefault", DictionaryOperations.DictionaryToAndroidJavaObject(DictionaryOperations.XMLtoDictionary(xmlPath)));
+
+        public void apply(ConfigValues values) => Call("apply", values);
+
+        public void clearAll() => Call("clearAll");
+
+        public void setDeveloperMode(bool isDeveloperMode) => Call("setDeveloperMode", isDeveloperMode);
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs:32]: Property

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        public Dictionary<String, System.Object> getMergedAll() => DictionaryOperations.AndroidJavaObjectToDictionary(Call<AndroidJavaObject>("getMergedAll"));
+
+        public ConfigValues loadLastFetched() => Call<ConfigValues>("loadLastFetched");
+
+        public Constants.SOURCE getSource(string key) => Call<Constants.SOURCE>("getSource", key);
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs:38]: Are you sure this is going to work? Returned value from Call is a Java enum.

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace HuaweiMobileServices.RemoteConfig
+{
+    public static class DEFAULT
+    {
+        public static bool BOOLEAN_VALUE = false;
+        public static double DOUBLE_VALUE = 0.0D;
+        public static long LONG_VALUE = 0L;
+        public static String STRING_VALUE = "";
+        public static byte[] BYTE_ARRAY_VALUE = new byte[0];
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs:11]: Don't hardcode values, use the values returned by the SDK.

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    public static class DictionaryOperations
+    {
+        private static readonly AndroidJavaClass sJavaArrays = new AndroidJavaClass("java.util.Arrays");
+
+        public static AndroidJavaObject DictionaryToAndroidJavaObject(IDictionary<String, System.Object> dictionary)
+        {
+            AndroidJavaObject javaMap = new AndroidJavaObject("java.util.HashMap");
+            IntPtr putMethod = AndroidJNIHelper.GetMethodID(
+                javaMap.GetRawClass(), "put",
+                    "(Ljava/lang/Object;Ljava/lang/Object;)Ljava/lang/Object;");
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:19]: Why this? Just use Call("put") like you do on AndroidJavaObjectToDictionary

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Xml;
+using System.Xml.Linq;
+using UnityEngine;
+
+namespace HuaweiMobileServices.Utils
+{
+    public static class DictionaryOperations
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:10]: Why public? This should be internal.

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        private static readonly AndroidJavaClass sJavaArrays = new AndroidJavaClass("java.util.Arrays");
+
+        public static AndroidJavaObject DictionaryToAndroidJavaObject(IDictionary<String, System.Object> dictionary)
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:14]: Internal not public
Transform to extension method and follow naming in other utils, like ToJavaHashMap()

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                        args[1] = v;
+                        AndroidJNI.CallObjectMethod(javaMap.GetRawObject(),
+                                putMethod, AndroidJNIHelper.CreateJNIArgArray(args));
+                    }
+                }
+            }
+            return javaMap;
+        }
+
+        public static Dictionary<String, System.Object> XMLtoDictionary(String xmlPath)
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:40]: What is this used for?

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+            foreach (var el in rootElement.Elements())
+            {
+                dictionary.Add(el.Attribute("key").Value, el.Value);
+            }
+
+            return dictionary;
+        }
+
+        public static Dictionary<String, System.Object> AndroidJavaObjectToDictionary(AndroidJavaObject javaList)
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:54]: Internal not public
Transform to extension method and follow naming in other utils, like ToDictionary().

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            if (javaList == null || javaList.Call<int>("size") == 0) return null;
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:56]: Return empty Dictionary instead of null

**Review by m0skit0 — CHANGES_REQUESTED** [review]

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 7 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:40]: with this function, i read xml file and convert it to dictionary

Comment (m0skit0) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:40]: Yes I understand, but why we need this? What XML are we reading?

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 8 above)
Comment (m0skit0) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:54]: Sorry, AsDictionary

**Comment by kayahanbaskeser** [comment]
> Please use C# naming conventions, methods start with uppercase, thanks.

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs (hunk 1 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs]: its done

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs (hunk 2 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/RemoteConfig/AGConnectConfigWrapper.cs]: you are right, this part did not work, I updated it

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs (hunk 3 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs]: its done

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 5 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs]: its done

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 6 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs]: it`s done

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 7 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs]: in remoteconfig, there is a method applydefault. and this method gets value from xml.

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 8 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs]: it`s done

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 9 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs]: it`s done

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs (hunk 4 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:17]: when i use call("put", "key", "value"), it gives error

Comment (m0skit0) [reply hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:17]: Which error, please be specific.

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs (hunk 3 above)
Comment (m0skit0) [reply hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs]: You moved DEFAULT inner class to AGConnectConfig?

File: hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                        args[1] = v;
+                        AndroidJNI.CallObjectMethod(javaMap.GetRawObject(),
+                                putMethod, AndroidJNIHelper.CreateJNIArgArray(args));
+                    }
+                }
+            }
+            return javaMap;
+        }
+
+        internal static Dictionary<string, object> XMLtoDictionary(String xmlPath)
```
Comment (m0skit0) [inline hms-sdk-unity/HuaweiMobileServices/Utils/DictionaryOperations.cs:38]: I still don't understand why we have this method, can you please explain? Where we use this?

**Review by m0skit0 — CHANGES_REQUESTED** [review]

File: hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs (hunk 3 above)
Comment (kayahanbaskeser) [reply hms-sdk-unity/HuaweiMobileServices/RemoteConfig/DEFAULT.cs]: I realized that it was the place on sdk

**Pull request merged**